### PR TITLE
Replaced `class Caller` with `struct Caller`. Internally a `CallerCon…

### PR DESCRIPTION
**Pull request by martindevans**
This PR replaces the Caller class with a Caller struct, removing a potential source of a _lot_ of allocation churn. The original idea (#208) was to use a `ref struct` to properly represent the true lifetime of a caller. However, that approach was very difficult to fit into the current system and would have required extensive rewrites. Instead this system throws an `ObjectDisposedException` if you try to use things beyond their lifetime (same as the old system).

The new API is very similar to the old one and should be mostly source compatible. Two breaking changes:
1. `Caller` no longer implements `IStore`. Internally we can call `AsStore()`, but externally it's simply no longer possible. I can make that `public` if you think it's necessary, but `IStore` seems like something that shouldn't really be used externally anyway.
2. `IDisposable` is implemented explicitly. Disposal is handled internally and calling `Dispose()` is not necessary. Implementing it explicitly hides it from intellisense, but means that any using statements around a `Caller` will continue to compile.

### Future Work

At the moment getting a `Memory` from a `Caller` will allocate a `CallerStore` and `Memory`. I plan to do followup PRs to provide a workaround for this but haven't decided on the best approach yet (for example I could add `Caller.GetMemorySpan(string name, long address, int length)`).

### CODE REVIEW

**Comment by kpreisser** [comment]
Hi,
note that with the discussion from [#200](https://github.com/bytecodealliance/wasmtime-dotnet/issues/200#issuecomment-1401167856) and PR #211 (if it's accepted), it would be possible to retrieve the originating `Store` instance from the `Caller`. This also removes the `IStore` interface.

With that changes, I think the `CallerStore` and `CallerContext` classes would no longer be required, as the `Store` can just be retrieved directly from the `Caller`.
What do you think?

Thanks!

**Comment by martindevans** [comment]
Storing the `Store` in the context is actually an approach I had considered for this work but I assumed the limited lifetime of objects retrieved from the `Caller` was there for a reason so I didn't want to change it.

The tricky bit of this work was tracking the lifetimes, that's why it needs the underlying `CallerContext` with an epoch number to track if it's referencing something that's disposed. However if we can directly get the `Store` there isn't any lifetime to track and the `Caller` really just becomes a thin wrapper around a `Store`. Immediate changes:
 - Remove `CallerStore` (use the main store)
 - Remove `CallerContext` (no context necessary)

Potential secondary changes:
 - `GetMemory`/`GetFunction` could have a cache in/alongside the `Store`, that way getting things in multiple calls only allocates the first time.

**Tl;dr: We should decide on #211 before reviewing this PR.**

**Comment by peterhuene** [comment]
Thanks, very much appreciated!

**Comment by peterhuene** [comment]
Closing this for the impending re-implementation (thanks for your patience on this @martindevans!).
